[Backport 3.6] PSA: use static key slots to store keys

valeriosetti opened the pull request:
## Description

This is the backport of #9302 to the 3.6 LTS branch. 

## PR checklist

- [x] **changelog** provided
- [x] **development PR** provided #9302 
- [x] **framework PR** not required
- [x] **3.6 PR** here
- [x] **2.28 PR** not required (new feature, no bug fix)
- **tests**  provided

valeriosetti commented:
I removed the `need-ci` label because the OpenCI is fine. Failures were only found in the InternalCI (which I cannot check), but since they are not confirmed by the OpenCI, I think they might be spurious failures.

tom-cosgrove-arm commented:
Note: conflicts

gilles-peskine-arm commented:
Now that the main PR has been approved, please complete and rebase this backport.

valeriosetti commented:
I've just backported all the commits from #9302 to `mbedtls-3.6`. Only some of them required a manual fix, in particular a couple of `*.data` files (pkwrite and pkparse) that were moved in `tf-psa-crypto` in the `development` branch.

tom-cosgrove-arm commented on tests/suites/test_suite_pkwrite.data:
```diff
@@ -7,12 +7,12 @@ depends_on:MBEDTLS_RSA_C
-depends_on:MBEDTLS_RSA_C:MBEDTLS_PEM_PARSE_C:MBEDTLS_PEM_WRITE_C
-pk_write_pubkey_check:"../framework/data_files/rsa4096_pub.pem":TEST_PEM
+depends_on:MBEDTLS_RSA_C:MBEDTLS_PEM_PARSE_C:MBEDTLS_PEM_WRITE_C:MBEDTLS_TEST_PK_ALLOW_RSA_KEY_PAIR_4096
+pk_write_pubkey_check:"../../framework/data_files/rsa4096_pub.pem":TEST_PEM
```
How come this line needs `../../framework` but line 7 only needs `../framework`?

valeriosetti replied:
Ops, sorry, that's one of the 2 files who had conflicts during the rebase and apprently I missed a change here :/

tom-cosgrove-arm reviewed (approved):
I've compared the diffs for each of the files in the two PRs - test_suite_pkparse.data had  by far the most changes - and I am happy that this is a proper backport of #9302

valeriosetti commented:
CI reported a lot of failures and I found that this was due to the `framework` not being properly updated. So:
- I rebased https://github.com/Mbed-TLS/mbedtls-framework/pull/48 on top of `framework` `main`
- I added a new (cherry-picked) commit to make use of the new `framework` reference

tom-cosgrove-arm reviewed (dismissed):
LGTM (subject to CI, of course)

gilles-peskine-arm reviewed (dismissed):
LGTM

This is a faithful backport of https://github.com/Mbed-TLS/mbedtls/pull/9302 and I can't think of anything that would need to be different. In particular, I am satisfied that there is no change in behavior, and no non-trivial change in code size, when the new option `MBEDTLS_PSA_STATIC_KEY_SLOTS` remains disabled.

gilles-peskine-arm reviewed (dismissed):
LGTM after rebase from 3558cba6504897e9f7dc4ecd5766c5cc2abff3e9 to update the framework

valeriosetti commented:
@bensze01 I'm adding you as reviewer for this one since it's the backport of #9302, where you are a reviewer, and it simply cherry-picks commits from there so it should be relatively fast to review.

valeriosetti commented:
I just backported the same changes done in `development` https://github.com/Mbed-TLS/mbedtls/pull/9302#issuecomment-2429027625

gilles-peskine-arm reviewed (approved):
Looks good to me after rebase from [my previous approval](https://github.com/Mbed-TLS/mbedtls/pull/9448#pullrequestreview-2372234717) at b287a4056d5639bb58743de924452a7d596a9f05 to the current head at 40859ac3b6ad62fccbd5078207480515b654bdbd.

tom-cosgrove-arm reviewed (approved):
LGTM
